mypy/constraints.py

```diff
@@ -339,6 +340,26 @@ def _infer_constraints(
     if isinstance(actual, AnyType) and actual.type_of_any == TypeOfAny.suggestion_engine:
         return []
 
+    # type[A | B] is always represented as type[A] | type[B] internally.
+    # This makes our constraint solver choke on type[T] <: type[A] | type[B],
+    # solving T as generic meet(A, B) which is often `object`. Force unwrap such unions
+    # if both sides are type[...] or unions thereof. See `testTypeVarType` test
+    def _is_type_type(tp: ProperType) -> TypeGuard[TypeType | UnionType]:
+        return (
+            isinstance(tp, TypeType)
+            or isinstance(tp, UnionType)
+            and all(isinstance(get_proper_type(o), TypeType) for o in tp.items)
+        )
+
+    def _unwrap_type_type(tp: TypeType | UnionType) -> ProperType:
+        if isinstance(tp, TypeType):
+            return tp.item
+        return UnionType.make_union([cast(TypeType, o).item for o in tp.items])
```

Comment:
I think the cast may fail if we have a type alias type. Adding `get_proper_type(o)` should help.